### PR TITLE
wasm binary parser

### PR DESCRIPTION
**Pull request by chfast**
wasm binary parser.

### CODE REVIEW

File: lib/fizzy/parser.cpp [hunk 1]
```diff
@@ -1,10 +1,29 @@
...
+    {
+        const auto id = static_cast<sectionid>(*it++);
+        const auto [size, new_pos] = leb128u_decode<uint32_t>(it);
+        it = new_pos;
+        switch (id)
+        {
+        default:
+            break;
+        }
+        it += size;
```
Comment (axic) [inline lib/fizzy/parser.cpp]: I think this should move into the `default` case.

Comment (chfast) [reply lib/fizzy/parser.cpp]: This is for all sections, also the "implemented" ones.

File: lib/fizzy/parser.hpp [hunk 2]
```diff
@@ -0,0 +1,72 @@
+#pragma once
+
+#include "types.hpp"
+#include <stdexcept>
+#include <tuple>
+
+namespace fizzy
+{
+constexpr uint8_t wasm_prefix_data[] = {0x00, 0x61, 0x73, 0xd6, 0x01, 0x00, 0x00, 0x00};
```
Comment (gumb0) [inline lib/fizzy/parser.hpp]: Does this need to be in a header?

Comment (chfast) [reply lib/fizzy/parser.hpp]: Used in tests.

File: lib/fizzy/parser.hpp [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        }
+    }
+};
+
+template <>
+struct parser<uint32_t>
+{
+    parser_result<uint32_t> operator()(const uint8_t* pos)
+    {
+        const auto b = *pos++;
```
Comment (gumb0) [inline lib/fizzy/parser.hpp]: should use `leb128u_decode<uint32_t>`

File: lib/fizzy/parser.cpp (hunk 1 above)
Comment (gumb0) [reply lib/fizzy/parser.cpp]: For parsed section we should check that parser consumed the expected number of bytes

File: lib/fizzy/parser.hpp (hunk 3 above)
Comment (chfast) [reply lib/fizzy/parser.hpp]: Use and simplified.

File: lib/fizzy/parser.cpp (hunk 1 above)
Comment (chfast) [reply lib/fizzy/parser.cpp]: Can it consume less?

Comment (chfast) [reply lib/fizzy/parser.cpp]: I will add the more restrictive check.

File: lib/fizzy/parser.cpp [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        case sectionid::type:
+            std::tie(mod.typesec, it) = parser<std::vector<functype>>{}(it);
+            break;
+        default:
+            it += size;
+            break;
+        }
+
+        if (it != expected_end_pos)
+            throw parser_error{"incorrected section " + std::to_string(static_cast<int>(id)) +
```
Comment (gumb0) [inline lib/fizzy/parser.cpp]:
```suggestion
            throw parser_error{"incorrect section " + std::to_string(static_cast<int>(id)) +
```

**Pull request merged**